Sources/PackageGraph/PackageGraph+Loading.swift

```diff
@@ -216,37 +216,76 @@ private func createResolvedPackages(
-                let resolvedPackage = packageMapByIdentity[PackageIdentity(url: url)]
-
-                // We check that the explicit package dependency name matches the package name.
-                if let resolvedPackage = resolvedPackage,
-                    let explicitDependencyName = dependency.explicitName,
-                    resolvedPackage.package.name != dependency.explicitName
-                {
-                    let error = PackageGraphError.incorrectPackageDependencyName(
-                        dependencyName: explicitDependencyName,
-                        dependencyURL: dependency.url,
-                        packageName: resolvedPackage.package.name)
+            // Otherwise, look it up by its identity.
+            if let resolvedPackage = packageMapByIdentity[dependencyIdentity] {
+                // check if this resolvedPackage already listed in the dependencies
+                // this means that the dependencies share the same identity
+                // FIXME: this works but the way we find out about this is based on a side effect, need to improve it when working on identity
+                guard !dependencies.contains(resolvedPackage) else {
+                    let error = PackageGraphError.dependencyAlreadySatisfiedByIdentifier(
+                        dependencyPackageName: package.name,
+                        dependencyURL: dependencyURL,
+                        otherDependencyURL: resolvedPackage.package.manifest.url,
+                        identity: dependencyIdentity)
                     let diagnosticLocation = PackageLocation.Local(name: package.name, packagePath: package.path)
-                    diagnostics.emit(error, location: diagnosticLocation)
+                    return diagnostics.emit(error, location: diagnosticLocation)
                 }
-
-                return resolvedPackage
+                // check that the explicit package dependency name matches the package name.
+                if let explicitDependencyName = dependency.explicitName, resolvedPackage.package.name != explicitDependencyName {
+                    // check if this resolvedPackage url is the same as the dependency one
+                    // if not, this means that the dependencies share the same identity
+                    // FIXME: this works but the way we find out about this is based on a side effect, need to improve it when working on identity
+                    if resolvedPackage.package.manifest.url != dependencyURL {
+                        let error = PackageGraphError.dependencyAlreadySatisfiedByIdentifier(
+                            dependencyPackageName: package.name,
+                            dependencyURL: dependencyURL,
+                            otherDependencyURL: resolvedPackage.package.manifest.url,
+                            identity: dependencyIdentity)
+                        let diagnosticLocation = PackageLocation.Local(name: package.name, packagePath: package.path)
```

Comment:
This is good for now, and in line with existing code.  I am preparing another a PR that fixes the diagnostic location in a couple of places, and as part of that, adds an extension on `Package` that constructs a diagnostic location that's either relative or remote as part of fixing a different issue.  So this is fine for now, and I will convert these locations to use that accessor as part of that PR.